File: Algorithms/Sorting/Heap_Sort/Heap_Sort.cs
using System; 
  
public class Heap_Sort 
{ 
    public void sort(int[] arr) 
    { 
        int n = arr.Length; 
  
        // Build heap 
        for (int i = n / 2 - 1; i >= 0; i--) 
            heapify(arr, n, i);

        for (int i=n-1; i>0; i--) 
        {  
            int temp = arr[0]; 
            arr[0] = arr[i]; 
            arr[i] = temp; 
            heapify(arr, i, 0); 
        } 
    } 
   
    void heapify(int[] arr, int n, int i) 
    { 
        int largest = i;  
        int l = 2*i + 1;  
        int r = 2*i + 2;  
   
        if (l < n && arr[l] > arr[largest]) 
            largest = l; 

        if (r < n && arr[r] > arr[largest]) 
            largest = r; 

        if (largest != i) 
        { 
            int swap = arr[i]; 
            arr[i] = arr[largest]; 
            arr[largest] = swap; 

            heapify(arr, n, largest); 
        } 
    } 

    static void printArray(int[] arr) 
    { 
        int n = arr.Length; 
        for (int i=0; i<n; ++i) 
            Console.Write(arr[i]+" "); 
        Console.Read(); 
    } 

    public static void Main() 
    { 
        Console.WriteLine("Enter array size : ");   
        int n = Convert.ToInt32(Console.ReadLine());
        int[] arr = new int[n]; 

        Console.WriteLine("Enter array elements : ");
        
        for(int i=0; i<n; i++){
            arr[i] = Convert.ToInt32(Console.ReadLine());
        }
        
        Heap_Sort ob = new Heap_Sort(); 
        ob.sort(arr); 
  
        Console.WriteLine("Sorted array is"); 
        printArray(arr); 
    } 
} 
  

File: Algorithms/Sorting/Heap_Sort/heap_sort.py
def heapify(arr, n, i):
    largest = i
    r = 2 * i + 2
    l = 2 * i + 1
    if l < n and arr[largest] < arr[l]:
        largest = l

    if r < n and arr[largest] < arr[r]:
        largest = r

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # swap
        heapify(arr, n, largest)


def heapSort(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # swap
        heapify(arr, i, 0)


arr = [120, 101, 13, 65, 2, 77]
heapSort(arr)
for i in arr:
    print(i)
